[BEAM-850] Adding display data to sink/sources PubSub, BigQuery, NativeFile, Fil…

pabloem opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:
- [x] Make sure the PR title is formatted like:
  `[BEAM-<Jira issue #>] Description of pull request`
- [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
     Travis-CI on your fork and ensure the whole test matrix passes).
- [x] Replace `<Jira issue #>` in the title with the actual Jira issue
     number, if there is one.
- [ ] If this contribution is large, please file an Apache
     [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).
---

…eBasedSource; Write, Read, ParDo transforms; DoFn.

robertwb commented on sdks/python/apache_beam/io/bigquery.py:
```diff
@@ -335,6 +336,24 @@ def __init__(self, table=None, dataset=None, project=None, query=None,
+  def display_data(self):
+    res = {}
+    if self.query is not None:
+      res['query'] = DisplayDataItem(self.query, label='Query')
```
What's the difference between label and key? Will users almost always have both? Do we need both?

pabloem replied:
*if users do not specify a label, the UI will show the key of the item.

swegner replied:
Phrased differently: 

* The namespace+key make up the URN for a display data item, useful for UI's to provide richer UI for well-known items (for example: recognize a GCS input path and provide a deep-link)
* The label is the human-readable name of the display item used in UI's. If not explicitly provided, defaults to the key.

pabloem commented:
r: @swegner since you added display data to a bunch of components, can you review my PR? : )

swegner commented on sdks/python/apache_beam/transforms/display.py:
```diff
@@ -86,13 +86,17 @@ def _populate_items(self, display_data_dict):
     """ Populates the list of display data items.
     """
     for key, element in display_data_dict.items():
+      if element is None:
```
In the Java SDK, we take a strict stance against registering null display data. Implicitly dropping null values in the SDK can lead to confusing behavior (_"I registered this display item, why isn't it showing up?"_).  Instead, the Java SDK has a helper method `.addIfNotNull(..)`, and the default `.add(..)` method throws if a null value is provided. This makes it very obvious whether you can reliably expect your display item to make it to the UI and when it might be dropped.

I recommend following this semantic here in the spirit of "element of least-surprise", but also so we are consist across SDKs.

pabloem replied:
That makes sense. I'll consult with Robert.

swegner commented on sdks/python/apache_beam/transforms/display_test.py:
```diff
@@ -22,8 +22,16 @@
+from apache_beam.transforms.display import *
+
+
+def make_nspace_display_data(component):
```
It looks like only a couple tests classes are using this-- should they all?

Note that the transforms and tests we have in the core SDK are likely to get copy-pasted many times over their life, for future SDK transforms and by users. Your implementation here establishes the recipe for how to implement and test display data.

swegner replied:
And on that note, I'll point out how having `DisplayDataMatchers` have been valuable in the Java SDK to lower the bar to writing concise display data tests. Some things that are difficult without them:

* Cannot easily assert on single part of display data. For example, if you have many tests of a PTransform that vary a single property to validate display data, each assertion will redundantly include the full set of display data items + properties. For example, see [`TextIOTest`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java#L538) `testWriteDisplayData()`, `testWriteDisplayDataValidateThenHeader()`, `testWriteDisplayDataValidateThenFooter()`.
* Tests for sub-component display data break encapsulation. For example, in order to test that the `Read` transform properly includes display data for its `Source`, your assert must list all display data for the `Source`. Later updating the `Source` display data will also break the `Read` transform test. In Java we have a `includesDisplayDataFrom(..)` matcher. See [`ReadTest.testDisplayData()`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/io/ReadTest.java#L99)
* Equality-tests on the display data map means adding additional elements to the display data schema (like `path`) will break all tests. It's convenient to assert on only the fields that a PTransform actually sets and exclude others filled in by the SDK (namespace and path).

swegner commented on sdks/python/apache_beam/transforms/ptransform_test.py:
```diff
@@ -663,6 +665,75 @@ def process(self, context):
+                                    namespace=nspace)
+    hc.assert_that(dd.items, hc.has_item(expected_item))
+
+    ########################
```
Each of these headings seems to be a separate test case.

pabloem replied:
Done.

swegner commented on sdks/python/setup.py:
```diff
@@ -93,6 +93,7 @@ def get_version():
     'protorpc>=0.9.1,<0.12',
     'python-gflags>=2.0,<4.0.0',
     'pyyaml>=3.10,<4.0.0',
+    'pyhamcrest>=1.9,<2.0',
```
@robertwb is this the list of dependencies that gets installed when submitting a Dataflow job? Do we want pyhamcrest here?

swegner commented on sdks/python/apache_beam/transforms/core.py:
```diff
@@ -29,7 +29,7 @@
 from apache_beam.internal import util
```
In Java the list of PTransforms with display data was much longer. Some components I didn't see here:

* Combine
* Window
* AvroIO
* XmlSource / XmlSink
* PipelineOptions

pabloem replied:
Yup. I'll work on those in a following transaction.

swegner commented on sdks/python/apache_beam/transforms/ptransform_test.py:
```diff
@@ -663,6 +665,75 @@ def process(self, context):
     self.check_label(beam.ParDo(MyDoFn()), r'ParDo(MyDoFn)')
 
 
+class PTransformTestDisplayData(unittest.TestCase):
```
In Java, we also wrote some "integration" tests to ensure that display data gets properly attached to a pipeline job. There are many ways this can fail that aren't captured in unit tests: 

* Transforms will often wrap or adapt their components, and must ensure that the display data is also properly wrapped.
* Runners can replace transforms in a pipeline graph and must provide equivalent display data in the replacement.
* Composite transforms are lots during pipeline translation and thus composite display data must be passed down into leaf transforms.

In Java SDK, we selected a small subset of "important" transforms to write display data integration tests for-- mostly IO connectors. See [`ReadTest.testBoundedPrimitiveDisplayData()`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/io/ReadTest.java#L110)

pabloem commented on sdks/python/apache_beam/io/bigquery.py:
```diff
@@ -335,6 +336,24 @@ def __init__(self, table=None, dataset=None, project=None, query=None,
     self.validate = validate
     self.coder = coder or RowAsDictJsonCoder()
 
+  def display_data(self):
```
This is a use case that I considered too. We could rely on the `mro()` method to obtain display data from super-classes - though we might not always want this (?). What do you think, @robertwb ?

robertwb replied:
I think it'd be good practice to call super (consistently).

pabloem commented on sdks/python/apache_beam/io/filebasedsource_test.py:
```diff
@@ -222,6 +225,15 @@ def test_fully_read_single_file(self):
+    nspace = '{}.{}'.format(fbs.__module__, fbs.__class__.__name__)
+    expected_items = [DisplayDataItem(file_name, key='filePattern',
+                                      label="File Pattern", namespace=nspace),
+                      DisplayDataItem('_CompressionType(auto)',
```
Done

pabloem commented on sdks/python/apache_beam/io/bigquery.py:
```diff
@@ -335,6 +336,24 @@ def __init__(self, table=None, dataset=None, project=None, query=None,
+                                   tableSpec)
+      res['table'] = DisplayDataItem(tableSpec, label='Table')
+
+    if not self.validate:
```
Done

pabloem commented on sdks/python/apache_beam/io/filebasedsource_test.py:
```diff
@@ -501,6 +513,16 @@ def test_source_creation_fails_for_non_number_offsets(self):
       SingleFileSource(
           fbs, file_name='dummy_file', start_offset=None, stop_offset=100)
 
+    dd = DisplayData.create_from(fbs)
```
Done

pabloem commented on sdks/python/apache_beam/io/iobase.py:
```diff
@@ -708,6 +714,10 @@ def __init__(self, *args, **kwargs):
 
+  def display_data(self):
+    return {'sink': self.sink.__class__,
+            'sink_dd': self.sink}
```
Done.

pabloem commented on sdks/python/apache_beam/io/pubsub.py:
```diff
@@ -54,6 +55,13 @@ def format(self):
+    return {'idLabel': DisplayDataItem(self.id_label,
+                                       label='ID Label Attribute'),
+            'topic': DisplayDataItem(self.topic, label='Pubsub Topic'),
+            'subscription': DisplayDataItem(self.subscription,
```
Done.

pabloem commented on sdks/python/apache_beam/io/fileio_test.py:
```diff
@@ -631,7 +660,16 @@ def test_write_text_bzip2_file(self):
-  def test_write_text_bzip2_file_auto(self):
+    nspace, dd = make_nspace_display_data(sink)
+    expected_items = [
+        DisplayDataItem('{}{}'.format(self.path, '-SSSSS-of-NNNNN'),
```
I looked at the code for the sources and couldn't find where the file name is created. It seems that it's just sent to the dataflow service. I think this should be tested in an integration test. - But I've added an extra test case to keep track of this in case it changes.

pabloem commented:
@swegner I have addressed your comments. Here's what I added and left for later:

* I **added** an `ItemMatcher` that helps simplify unit tests for display data. I have also added the `ItemMatcher` to our current tests.
* I **added** functions `DisplayDataItem.drop_if_none()`, and `DisplayDataItem.drop_if_default(def_val)` to be consistent with the model in the Java SDK.
* I **added** an extra text case for file pattern prefix/suffix treatment in `NativeFileSink`. We have only two cases at the moment.
* I did **not add** exhaustive test cases for file pattern prefix/suffix treatment in `NativeFileSink`. The code for that seems to not be part of the Python SDK.
* I did **not add** small "integration tests" to make sure display data gets added - I'd like to leave this for a later change (after I've thoroughly studied Java's [`ReadTest.testBoundedPrimitiveDisplayData()`](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/io/ReadTest.java#L110)).

Please review, and let me know what you think.
-P. : )

swegner commented on sdks/python/apache_beam/io/bigquery.py:
```diff
@@ -335,6 +336,23 @@ def __init__(self, table=None, dataset=None, project=None, query=None,
+  def display_data(self):
+    res = {}
+    if self.query is not None:
+      res['query'] = DisplayDataItem(self.query, label='Query')
```
Convert to use `.drop_if_none()`, to establish the convention.

pabloem replied:
Done.

swegner commented on sdks/python/apache_beam/transforms/display_test.py:
```diff
@@ -22,12 +22,76 @@
 from apache_beam.transforms.display import DisplayDataItem
 
 
+def make_nspace_display_data(component):
```
pydoc

swegner commented on sdks/python/apache_beam/transforms/display_test.py:
```diff
@@ -22,12 +22,76 @@
+  return nspace, dd
+
+
+class ItemMatcher(BaseMatcher):
```
pydoc class, and public methods

pabloem replied:
Done. (`_matches`, and `describe_to` methods come from the super class.)

swegner commented on sdks/python/apache_beam/transforms/display_test.py:
```diff
@@ -22,12 +22,76 @@
+def make_nspace_display_data(component):
+  dd = DisplayData.create_from(component)
+  nspace = '{}.{}'.format(component.__module__, component.__class__.__name__)
+  return nspace, dd
```
The usage of this method for assertions feels strange because half of the output (`nspace`) feeds into the 'expected' display data, and the other half (`dd`) represents the 'actual'. It doesn't seem intuitive-- without reading the implementation I might expect the return `nspace` to be pulled from the generated display data and represent the 'actual nspace'.

Is it necessary to validate `nspace` on each transform display data test? Since this is generated by the SDK rather than from inputs I think we should only need one test in the display data unit tests and then disregard it for other validations.

pabloem replied:
Makes sense. I'm on it.

swegner commented on sdks/python/apache_beam/io/bigquery_test.py:
```diff
@@ -111,6 +114,39 @@ def test_invalid_json_neg_inf(self):
+
+    nspace, dd = make_nspace_display_data(source)
+    expected_items = [
+        ItemMatcher.matches_kvn('validation', True, nspace),
```
Display data tests tend to live among other tests within a component test class. As such, the imported `ItemMatcher` class isn't obviously display-data-specific in context. Is there a better name that makes it contextually obvious that `ItemMatcher` is specific to display data?  Perhaps `DisplayDataMatcher.item_matches(..)`?

swegner commented on sdks/python/apache_beam/transforms/core.py:
```diff
@@ -580,6 +590,12 @@ def make_fn(self, fn):
     return self.fn.process_argspec_fn()
 
+  def display_data(self):
+    # TODO: Add special handling of callables
```
Are these TODO's still relevant?

pabloem replied:
I removed that TODO.

swegner commented on sdks/python/apache_beam/transforms/display_test.py:
```diff
@@ -22,12 +22,76 @@
+      description.append('namespace is {} '.format(self.namespace))
+
+  @classmethod
+  def matches_key(cls, key):
```
So far I only see usage of `matches_kvn`. I recommend removing the additional APIs you aren't using-- it's easier to start with a small API and grow it later than to start with a wide API and narrow it.

swegner commented on sdks/python/apache_beam/transforms/display.py:
```diff
@@ -145,6 +148,40 @@ def __init__(self, value, url=None, label=None,
+    self._drop_if_none = False
+    self._drop_if_default = False
+
+  def drop_if_none(self):
```
Add unit tests for `drop_if_none()` and `drop_if_default()`

pabloem replied:
Done.

swegner commented on sdks/python/apache_beam/transforms/ptransform_test.py:
```diff
@@ -663,6 +665,80 @@ def process(self, context):
 
+class PTransformTestDisplayData(unittest.TestCase):
+  def test_map_named_function(self):
+    ########################
```
Now that these are separate tests, do these block comments add any value? If not, remove.

pabloem replied:
Done.

swegner commented on sdks/python/tox.ini:
```diff
@@ -27,6 +27,7 @@ select = E3
 deps=
   pep8
   pylint
+  pyhamcrest
```
Now that this `pyhamcrest` is added here, should it be removed from `setup.py`?

swegner replied:
My understanding of `setup.py` is that it defines how the Beam SDK gets installed, and thus also  dictates the list of dependencies installed on worker VMs to execute Beam pipelines. If that's the case, we should probably keep the dependency list as trim as possible.

@robertwb / @chamikaramj, can you shed light on whether this is the proper way to add a test dependency on `pyhamcrest`?

chamikaramj replied:
Probably we should keep this dependency in setup.py as well to support 'python setup.py test' command. @charlesccychen to confirm.

charlesccychen replied:
I suggest keeping it in both places and adding this to test_requires in setup.py.  Unfortunately, this doesn't interact well with pip (https://github.com/pypa/pip/issues/1197), but semantically, that is the correct place to put it and it will make python setup.py test work.

robertwb replied:
Agreed.

pabloem commented:
I have addressed all comments, and unit tests are passing; but I am having jobs fail on the Dataflow service with these changes. : (
I'm looking into it.

charlesccychen commented on sdks/python/setup.py:
```diff
@@ -93,6 +93,10 @@ def get_version():
+    'pyhamcrest>=1.9,<2.0',
+    ]
+REQUIRED_TEST_PACKAGES = [
+    'pyhamcrest>=1.9,<2.0',
```
I suggest that we keep this pyhamcrest and remove the one in REQUIRED_PACKAGES above.

pabloem replied:
Done.

charlesccychen commented on sdks/python/apache_beam/io/filebasedsource_test.py:
```diff
@@ -482,8 +496,8 @@ def setUp(self):
     stop_not_a_number_error = 'stop_offset must be a number*'
-
-    fbs = LineSource('dymmy_pattern')
+    file_name = 'dymmy_pattern'
```
driveby nit: "dymmy_pattern" is misspelled throughout.

swegner commented on sdks/python/apache_beam/transforms/display_test.py:
```diff
@@ -137,14 +124,39 @@ def display_data(self):
+                                                        'something'),
+                      DisplayDataItemMatcher.matches_kv('nodef_val',
+                                                        True)]
+    hc.assert_that(dd.items, hc.contains_inanyorder(*expected_items))
```
Does this validate that the `dd.items` contains *only* the specified items?  You might need to add an additional assertion `hc.assert_that(dd.items, hc.not(hc.contains_inanyorder(..`

pabloem replied:
`contains_inanyorder` [should match the entire sequence](https://pypi.python.org/pypi/PyHamcrest). But just in case, I added the check. : )

robertwb replied:
The check is redundant, better to omit it.

pabloem replied:
Okay. Done.

pabloem commented:
Merging with HEAD of `python-sdk` branch fixed issue running jobs on Dataflow. @robertwb can you take a look and merge if appropriate? : )

robertwb commented on sdks/python/apache_beam/io/bigquery.py:
```diff
@@ -335,6 +336,22 @@ def __init__(self, table=None, dataset=None, project=None, query=None,
+                                         label='Validation Enabled')}
+
+    if self.table_reference is not None:
+      tableSpec = '{}.{}'.format(self.table_reference.datasetId,
```
Put this in an else clause of the if below.

pabloem replied:
Done.

robertwb commented on sdks/python/apache_beam/io/bigquery.py:
```diff
@@ -335,6 +336,22 @@ def __init__(self, table=None, dataset=None, project=None, query=None,
     self.coder = coder or RowAsDictJsonCoder()
 
+  def display_data(self):
+    res = {'query':
```
Is query mutually exclusive with table_reference?

pabloem replied:
Yes . If both are set or both are unset, the constructor raises `ValueError`.

robertwb replied:
In that case, I'd put it in the else clause to make things clearer.

robertwb commented on sdks/python/apache_beam/io/bigquery_test.py:
```diff
@@ -134,6 +178,15 @@ def test_specify_query_without_table(self):
 
+  def test_schema_descriptor_display_data(self):
+    sink = beam.io.BigQuerySink(
+        'dataset.table', schema='s:STRING, n:INTEGER')
```
Doesn't seem to test the schema.

pabloem replied:
The schema is not display data in this case. I've removed it in the test.

robertwb commented on sdks/python/apache_beam/io/bigquery_test.py:
```diff
@@ -143,9 +196,18 @@ def test_parse_schema_descriptor(self):
         field.name: field.type for field in sink.table_schema.fields}
     self.assertEqual({'n': 'INTEGER', 's': 'STRING'}, result_schema)
 
+  def test_project_table_schema_display_data(self):
```
same

pabloem replied:
Same ; )

robertwb commented on sdks/python/apache_beam/io/fileio_test.py:
```diff
@@ -592,6 +602,41 @@ def test_write_text_file(self):
+    hc.assert_that(dd.items, hc.contains_inanyorder(*expected_items))
+
+  def test_text_file_display_data_added_suffix(self):
+    sink = fileio.NativeTextFileSink('{}.{}'.format(self.path, 'pdf'))
```
This test seems to be inconsistent whether the period belongs to the extension or the format pattern.

pabloem replied:
In this case, the period is part of the pattern. I added it for the future, in case we decide to be smart about parsing the file extensions. The next test (`test_text_file_display_data_suffix`) actually provides a pattern+suffix.

robertwb replied:
OK, what is this testing in that case? I'd leave the pdf off if it's not really part of the test, as it confuses things.

robertwb commented on sdks/python/apache_beam/transforms/display_test.py:
```diff
@@ -72,50 +123,56 @@ def display_data(self):
-        self.fn = fn
-
+    expected_items = [
+        DisplayDataItemMatcher.matches_kvn('complex_url',
```
Aren't these tests now weaker, as they don't test the short values or labels?

pabloem replied:
You're right. I've added the matchers for this.

robertwb commented on sdks/python/apache_beam/transforms/display_test.py:
```diff
@@ -78,6 +89,16 @@ def matches_kvn(cls, key, value, namespace):
                match_props=['key', 'value', 'namespace'])
 
+  @classmethod
+  def matches_all(cls, key, value, namespace, label, shortValue):
```
Using match_props and adding a new method for every subset doesn't seem very extensible. What if instead you just had

```
class DisplayDataItemMatcher(...):
  IGNORED = object()
  def __init__(self, key, value=IGNORED, namespace=IGNORED, ...)
    ...

```

and so DisplayDataItemMatcher(...) simply matched the keywords passed.

pabloem replied:
Done.

robertwb commented on sdks/python/apache_beam/io/fileio_test.py:
```diff
@@ -631,7 +689,20 @@ def test_write_text_bzip2_file(self):
+            'bzip2')]
+    hc.assert_that(dd.items, hc.contains_inanyorder(*expected_items))
+
+  def test_write_text_zlib_file_auto(self):
```
zlib?

robertwb commented:
LGTM, but doesn't seem to rebase cleanly on https://github.com/apache/incubator-beam/commit/ce3f1726ff9c69d718c41cae02cf2d4a370f5a22
